### PR TITLE
[CSV Sniffer] Selecting file to sniff from Glob and List

### PR DESCRIPTION
**Pull request by pdet**
This PR implements an adjustment to the CSV Sniffer regarding selecting the file used for sniffing.

Note that the column names and types depend on the file used for sniffing.

The new rule checks if the read_csv function is given a list of files with a `size > 1`. If so, and if the first file is not a glob pattern, that file will be used as the sniffed file for the function.

For example, in all the cases below, a.csv would be used as the sniffed file:

For example, all options below would use `a.csv` as the sniffed file:
* `read_csv(['a.csv'])`
* `read_csv(['a.csv', 'b_1.csv', 'b_2. csv'])`
* `read_csv(['a.csv', 'b_*.csv'])`

However, if the first path is a glob, we search the first 10 files for either a file larger than 1 MB or the largest file and use it for sniffing.

For example, the following options would trigger the search algorithm:

* `read_csv(['b_*.csv'])`
* `read_csv(['b_*.csv', 'a.csv'])`

Fix: https://github.com/duckdblabs/duckdb-internal/issues/1824
Related to: https://github.com/duckdb/duckdb/discussions/11909

### CODE REVIEW

**Comment by pdet** [comment]
@Mytherin I've retagged this to feature.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! It would be nicer to do this during the sniffing process itself by looking at how many rows were sniffed from each file. If we sniffed too few rows, we could open up a new file and continue the sniffing process (merging schemas as we go). This would avoid the extra file opening/closing we're doing now in `FindFileToSniff`, and would generally be more robust.

File: src/include/duckdb/common/multi_file_list.hpp [hunk 1]
```diff
@@ -157,6 +163,9 @@ class GlobMultiFileList : public MultiFileList {
 	vector<string> GetAllFiles() override;
 	FileExpandResult GetExpandResult() override;
 	idx_t GetTotalFileCount() override;
+	void SwapToFirst(const idx_t file_idx) override;
```
Comment (Mytherin) [inline src/include/duckdb/common/multi_file_list.hpp:166]: Are these changes still necessary?

File: src/function/table/read_csv.cpp [hunk 2]
```diff
@@ -44,6 +45,78 @@ void ReadCSVData::FinalizeRead(ClientContext &context) {
...
+		idx_t rows_read = sniffer.LinesSniffed() -
+		                  (options.dialect_options.skip_rows.GetValue() + options.dialect_options.header.GetValue());
+
+		schemas.emplace_back(sniffer_result.names, sniffer_result.return_types, file_paths[0], rows_read,
+		                     result.buffer_manager->GetBuffer(0)->actual_size == 0);
+		total_number_of_rows += sniffer.LinesSniffed();
+	}
+
+	// We do a copy of the options to not pollute the options of the first file.
+	while (total_number_of_rows < required_number_of_lines && current_file + 1 < file_paths.size()) {
```
Comment (Mytherin) [inline src/function/table/read_csv.cpp:77]: Should we add some upper-bound to how many files we read here, e.g. 10 files or so?

**Review by Mytherin — COMMENTED** [review]
Thanks for the fixes! Looks great - some more minor comments:

**Comment by Mytherin** [comment]
Thanks!
